## Import libraries and datasets

In [62]:
# import libraries
import numpy as np
import pandas as pd
import os

In [91]:
# import data frames 'orders' & 'products'
path = r'C:\Users\98912\MAY 2022 Instacart Analysis'

df_ords = pd.read_csv(os.path.join(path, '02-Data', 'Original Data', 'orders.csv'), index_col=False)
df_prods = pd.read_csv(os.path.join(path, '02-Data', 'Original Data', 'products.csv'), index_col=False)

## Data quality check

In [64]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

## Test data frame- checking mixed data type

In [65]:
df_test = pd.DataFrame()
df_test['mix'] = ['a', 'b', 1, True]
df_test

mix
0     a
1     b
2     1
3  True

In [66]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [72]:
for col in df_test.columns.tolist():
    if len(set((df_test[['mix']].applymap(type))['mix'].tolist())) > 1:
        print (col)

mix


## Missing value

In [73]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [74]:
df_nan = df_prods[df_prods['product_name'].isnull()==True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [75]:
df_prods.shape

(49693, 5)

In [76]:
df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

In [77]:
df_prods_clean.shape

(49677, 5)

## Duplicate

In [78]:
df_prods_dup = df_prods_clean[df_prods_clean.duplicated()]
df_prods_dup

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [79]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
df_prods_clean_no_dups.shape

(49672, 5)

In [80]:
# export the final dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02-Data', 'Prepared Data', 'products_checked.csv'))

 ## Task 4.5

In [81]:
# part 2: Run describe for df_prods
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#### First of all the maximum number for 'product_id' is 49688 while the count of rows in this table is 49693 it means that 5 rows have the same product_id or doesn't have any product_id at all. As product_id is the primary key in this table and every product should have id and this id must be unique, it seems that we have 5 duplicated rows. and it has been confirmed by previous checks in the exercise. 
<br>

#### the other noticable point the maximum number for proces which is 99999! This price seems to be high for the stuffs sold by instacart. Moreover the mean of price is 11.2 which shows that 99999 is probably an outlier in this column. Therefore, we need to investigate further the items that have this price. Maybe this is a wrong number or just an indicator for null values. 

In [82]:
# part3 : check for mixed data type in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col + ' contains mixed data types')
  else:
    print (col + ' doesn\'t contain mixed data types')


order_id doesn't contain mixed data types
user_id doesn't contain mixed data types
eval_set doesn't contain mixed data types
order_number doesn't contain mixed data types
order_dow doesn't contain mixed data types
order_hour_of_day doesn't contain mixed data types
days_since_prior_order doesn't contain mixed data types


In [83]:
# part 4: there is no column with mixed data type

In [84]:
# part 5: finding missing values in df_ords
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [85]:
df_ords_null = df_ords[df_ords['days_since_prior_order'].isnull()==True]
df_ords_null

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [86]:
df_ords_null[df_ords_null.duplicated()]

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

#### The interesting point about the number of null values (206209) is that this is exactly the number of user_id in 'orders' table and when we had a look in the df_ords_null we realized that there is no duplicated rows. Therefore, we can conclude that this null value came out for every user_id and has happend just once. As a result, it can be probably the first order of each user_id, and as there was not a previous record for them this value of 'days_since_prior_order' is considered or calculated as 'null'. 

In [87]:
df_ords[df_ords['user_id'] == 1]

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   1187899        1    train            11          4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

#### part 6: As the information related to first orders are valuable and useful for further analysis we cannot just simply delet the rows. So we have two options: 1- to build a new variable to indicate that this row is the first order of the user. 2- impute the null with an appropriate value. The first option I believe is not necessary, as we have the variable 'order_number' that shows the sequence number of orders and it shows number '1' for all rows with null in 'days_since_prior_order'. So I've decided to impute the null with '0' as it is the most logical value in this case.

In [92]:
df_ords_clean = df_ords.copy()
df_ords_clean['days_since_prior_order'].fillna(0, inplace=True) 
df_ords_clean.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     0.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [93]:
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [96]:
# part 7: check for duplicates in df_ords_clean
df_ords_dup = df_ords_clean[df_ords_clean.duplicated()==True]
df_ords_dup

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

#### There is no duplicated values in this dataframe

#### part 8: There is no need to address duplicated values

In [98]:
# as there was not any changes in df_prods_clean_no_dups and we have exportes it before, 
# we just export the new version of df_ords_clean
df_ords_clean.to_csv(os.path.join(path, '02-Data', 'Prepared Data', 'orders_checked.csv'))